README

unrar is a free software version of the non-free unrar utility that exists
on many GNU/Linux systems today. It uses the GPL'd UniquE RAR Library by
Christian Scheurer and Johannes Winkelmann. This program is a simple
command-line front-end to this library, and can list and extract RAR
archives. It does not rival the non-free unrar in terms of features, but
special care has been taken to ensure it meets most user's needs and it is
hoped that GUI archivers such as GNOME's (file-roller) and KDE's ark will
use this utility as a back-end.

	Usage: unrar [OPTION...] ARCHIVE [FILE...] [DESTINATION]
	Extract files from rar archives.
 
	  -x, --extract              Extract files from archive (default)
	  -t, --list                 List files in archive
	  -f, --force                Overwrite files when extracting
	      --extract-newer        Only extract newer files from the archive
	      --extract-no-paths     Don't create directories while extracting
	  -p, --password             Decrypt archive using a password
	  -?, --help                 Give this help list
	      --usage                Give a short usage message
	  -V, --version              Print program version
 
	Report bugs to <benasselstine@canada.com>.


You can download the latest version from http://gna.org/files/?group=unrar

[1] GPL'd UniquE RAR Library, http://www.unrarlib.org/
[2] file-roller, http://fileroller.sourceforge.net/



ChangeLog
2004-06-14  Ben Asselstine  <ben@asselstine.com>

	* src/unrarlib.h, src/Makefile.am: Moved the _DEBUG_LOG define into 
	Makefile.am, and commented it out.
	* src/opts.c (parse_opts): fixed calling of getpass, so it only gets
	called once even if --password is on the command-line twice.
	* src/opts.c (parse_opts): now using realloc(n,NULL) to behave like
	malloc(n), so that memory allocation only happens in one place.
	* src/Makefile.in: regenerated.
	* src/unrar.c: added Jeroen's name to the visible copyright notice.

2004-06-05  Jeroen Dekkers  <jeroen@dekkers.cx>

	* src/opts.c (options): Put string constants directly in the
	structure.




src/opts.h
/* Copyright (C) 2004  Jeroen Dekkers <jeroen@dekkers.cx>
   Copyright (C) 2004  Ben Asselstine <benasselstine@canada.com>

   This program is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License
   as published by the Free Software Foundation; either version 2
   of the License, or (at your option) any later version.
   
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License along
   with this program; if not, write to the Free Software Foundation, Inc.,
   51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA */

#ifndef FUNRAR_OPTS_H
#define FUNRAR_OPTS_H 1

enum stow_options_enum_t
{
  OPT_NONE,
  OPT_EXTRACT_NEWER,
  OPT_JUNK_PATHS,
  OPT_FORCE = 'f',
  OPT_PASSWORD = 'p',
  CMD_LIST = 't',
  OPT_VERBOSE = 'v',
  CMD_EXTRACT = 'x',
};
enum unrar_mode_enum_t
{
  MODE_UNKNOWN,
  MODE_EXTRACT,
  MODE_LIST,
};
struct unrar_arguments_t
{
  int mode;			//stow_mode_enum_t
  int force;
  int junk_paths;
  int extract_newer;
  char *destination_dir;
  char *archive_filename;
  int verbose;
  char *password;
};

struct arguments_t
{
  char **args;
  int arraylen;			//for argument processing
  struct unrar_arguments_t unrar;
};
int parse_opts (int argc, char **argv, struct arguments_t *arguments);
int compat_parse_opts (int argc, char **argv, struct arguments_t *arguments);

#endif


src/unrar.c
/* Copyright (C) 2004  Jeroen Dekkers <jeroen@dekkers.cx>
   Copyright (C) 2004  Ben Asselstine <benasselstine@canada.com>

   This program is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License
   as published by the Free Software Foundation; either version 2
   of the License, or (at your option) any later version.
   
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License along
   with this program; if not, write to the Free Software Foundation, Inc.,
   51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <error.h>
#include <errno.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <fcntl.h>
#include <unistd.h>
#include "opts.h"
#include "unrar.h"
#include "unrarlib.h"

/*
 *
 * dos_to_unix_time came from unzip sources.
 * dump_file is based on a simple GPL'd unrar client by Andreas F Borchert.
 * 
 *
 */
extern struct arguments_t arguments;

void
show_copyright ()
{
  printf ("\n");
  printf ("%s %s  Copyright (C) 2004  Ben Asselstine, Jeroen Dekkers\n", 
		  PACKAGE, VERSION);
  printf ("\n");
  return;
}

void
show_list_header (struct unrar_arguments_t *unrar)
{
  printf ("\n");
  printf ("RAR archive %s\n", unrar->archive_filename);
  printf ("\n");
  printf ("Pathname/Comment\n");
  printf
    ("                  Size   Packed Ratio  Date   Time     Attr      CRC   Meth Ver\n");
  printf
    ("-------------------------------------------------------------------------------\n");
  return;
}

void
show_list_footer (struct unrar_arguments_t *unrar)
{
  printf
    ("-------------------------------------------------------------------------------\n");
  return;
}

void
show_list_stats (unsigned int num_files, unsigned int num_compressed_bytes,
		 unsigned int num_bytes)
{
  float ratio;
  ratio = ((float) num_compressed_bytes / (float) num_bytes) * 100;
  printf ("%5u       %10u %8u %3.0f%%\n", num_files, num_bytes,
	  num_compressed_bytes, ratio);
  return;
}

#define DOSTIME_2038_01_18 ((unsigned long)0x74320000L)
#define U_TIME_T_MAX  ((time_t)(unsigned long)0xffffffffL)
#define S_TIME_T_MAX  ((time_t)(unsigned long)0x7fffffffL)
#define YRBASE  1900

time_t
dos_to_unix_time (unsigned long dosdatetime)
{
  time_t m_time;
  struct tm *tm;
  time_t now = time (NULL);

  tm = localtime (&now);
  tm->tm_isdst = -1;		/* let mktime determine if DST is in effect */

  /* dissect date */
  tm->tm_year = ((int) (dosdatetime >> 25) & 0x7f) + (1980 - YRBASE);
  tm->tm_mon = ((int) (dosdatetime >> 21) & 0x0f) - 1;
  tm->tm_mday = ((int) (dosdatetime >> 16) & 0x1f);

  /* dissect time */
  tm->tm_hour = (int) ((unsigned) dosdatetime >> 11) & 0x1f;
  tm->tm_min = (int) ((unsigned) dosdatetime >> 5) & 0x3f;
  tm->tm_sec = (int) ((unsigned) dosdatetime << 1) & 0x3e;

  m_time = mktime (tm);

  if ((dosdatetime >= DOSTIME_2038_01_18) && (m_time < (time_t) 0x70000000L))
    m_time = U_TIME_T_MAX;	/* saturate in case of (unsigned) overflow */
  if (m_time < (time_t) 0L)	/* a converted DOS time cannot be negative */
    m_time = S_TIME_T_MAX;	/*  -> saturate at max signed time_t value */
  return m_time;

}				/* end function dos_to_unix_time() */

void
replace_backslash_with_slash (char *s1)
{
  char *ptr;
  if (!s1)
    return;
  while ((ptr = strchr (s1, '\\')))
    {
      ptr[0] = '/';
      s1 = &ptr[1];
    }
  return;
}

void
unrar_list_item (struct unrar_arguments_t *unrar,
		 ArchiveList_struct * listptr)
{
  char datetime[20];
  char *attr;
  float ratio;

  printf (" %s\n", listptr->item.Name);
  if (listptr->item.UnpSize == 0)
    ratio = 0;
  else
    ratio =
      ((float) listptr->item.PackSize / (float) listptr->item.UnpSize) * 100;
  if (ratio > 100.0)
    ratio = 100.0;
  if (listptr->item.FileTime)
    {
      struct tm filetime;
      time_t unixtime;
      unixtime = dos_to_unix_time ((unsigned long) listptr->item.FileTime);
      localtime_r ((const time_t *) &unixtime, &filetime);
      strftime (datetime, sizeof (datetime), "%d-%m-%y %H:%M", &filetime);
    }
  else
    snprintf (datetime, sizeof (datetime), "00-00-00 00:00");

  if (listptr->item.FileAttr & 0x10)
    attr = ".D....";
  else
    attr = ".....A";

  printf ("            %10u %8u %3.0f%% %s   %s   %08X m%c? %3.1f\n",
	  (unsigned int) listptr->item.UnpSize,
	  (unsigned int) listptr->item.PackSize,
	  ratio,
	  datetime,
	  attr,
	  (unsigned int) listptr->item.FileCRC,
	  listptr->item.Method, (float) (listptr->item.UnpVer / 10.0));
  return;
}

int
unrar_list (struct unrar_arguments_t *unrar, int num_files, char **files)
{
  ArchiveList_struct list;
  ArchiveList_struct *listptr;
  int i, j, n;
  unsigned int count_files = 0;
  unsigned int count_bytes = 0;
  unsigned int count_compressed_bytes = 0;
  list.next = NULL;
  
  n = urarlib_list (unrar->archive_filename,
		    &list.next);
  if (n < 0)
    return n;
  if (unrar->verbose)
    printf ("showing %d files...\n", n);
  if (unrar->verbose)
    printf ("given a list of %d files.\n", num_files);
  listptr = list.next;
  show_list_header (unrar);
  for (i = 0; i < n; i++, listptr = listptr->next)
    {
      replace_backslash_with_slash (listptr->item.Name);
      if (num_files)
	{
	  int found = 0;
	  for (j = 0; j < num_files; j++)
	    {
	      if (unrar->verbose)
		printf ("comparing '%s' vs '%s'\n", files[j],
			listptr->item.Name);
	      if (strcmp (files[j], listptr->item.Name) == 0)
		{
		  found = 1;
		  break;
		}

	    }
	  if (!found)
	    continue;
	}
      count_files++;
      count_compressed_bytes += listptr->item.PackSize;
      count_bytes += listptr->item.UnpSize;
      unrar_list_item (unrar, listptr);
      if (listptr == NULL)	//sanity check.
	break;
    }

  urarlib_freelist (list.next);
  show_list_footer (unrar);
  show_list_stats (count_files, count_compressed_bytes, count_bytes);

  return 0;
}

int
unrar_mkpath (char *path) 
{
  char *path_tmp;
  int i,ret=0;
  if (path == NULL || path[0] == '\0') return (-1);
  path_tmp = strdup(path);
  if (path_tmp == NULL) return (-1);
  for (i=0; path_tmp[i] != '\0'; i++) {
    if (path_tmp[i]=='/') {
      path_tmp[i] = '\0';
      if (path_tmp[0] != '\0') {
        ret = mkdir(path_tmp,0777);
      }
      path_tmp[i] = '/';
    }
  }
  free(path_tmp);
  return ret;
}

int
dump_file (struct unrar_arguments_t *unrar, char *filename, void *data,
	   unsigned long data_size)
{
  int fd;
  int written;
  ssize_t nbytes = 0;
  int fcntl_flags = O_WRONLY | O_CREAT;
  if (unrar->force)
    fcntl_flags |= O_TRUNC;
  else
    fcntl_flags |= O_EXCL;
  fd = open (filename, fcntl_flags, 0666);
  if (fd < 0 && errno == ENOENT) {
    unrar_mkpath (filename);
    fd = open (filename, fcntl_flags, 0666);
  }
  if (fd < 0)
    return -1;

  while (nbytes < data_size)
    {
      written = write (fd, data + nbytes, data_size - nbytes);
      if (written < 0)
	return -1;
      nbytes += written;
    }
  if (close (fd) < 0)
    return -1;
  return 0;
}

void
show_status_line (char *action, char *file, char *status)
{
  printf ("%-11s %-57s %-10s\n", action, file, status);
  return;
}

int
unrar_extract_directory (struct unrar_arguments_t *unrar, char *dir)
{
  char *destination = NULL;
  int ret=0;
  if (unrar->junk_paths)
    return 0;
  if (asprintf (&destination, "%s/%s/", unrar->destination_dir, dir) == -1)
    {
      error (0, 0, "asprintf failed: %m\n");
      return -1;
    }
  if (mkdir (destination, 0777) < 0)
    {
      switch (errno)
	{
	  case EEXIST:
	    if (unrar->force)
	      chmod (destination, 0777);
            break;
          case ENOENT:
            ret = unrar_mkpath(destination);
            if (ret==(-1))
              {
                error(0, 0, "mkpath failed '%s': %m\n", destination);
              }
            break;
          default:
            error (0, 0, "mkdir failed '%s': %m\n", destination);
            ret=(-1);
            break;
	}
    }
  free (destination);
  return ret;
}

int
unrar_extract_file (struct unrar_arguments_t *unrar, char *filename,
		    char *archive_member, time_t FileTime)
{
  int retval;
  char *destination = NULL;
  void *data = NULL;
  unsigned long data_size;
  if (!urarlib_get
      (&data, &data_size, archive_member, unrar->archive_filename,
       unrar->password))
    return -1;
  else
    {
      char *file;
      if (unrar->junk_paths)
	{
	  file = strrchr (filename, '/');
	  if (!file)
	    file = filename;
	}
      else
	file = filename;

      if (asprintf (&destination, "%s/%s", unrar->destination_dir, file) ==
	  -1)
	{
	  error (0, 0, "asprintf failed: %m\n");
	  return 0;
	}
      //where do i put it?  in destination.
      if (unrar->extract_newer)
	{
	  struct stat statbuf;
	  time_t unixtime;
	  int non_existent_files_are_not_newer = 1;
	  retval = lstat (destination, &statbuf);
	  if (retval < 0)
	    {
	      if (non_existent_files_are_not_newer)
		{
		  free (destination);
		  return 1;
		}
	    }
	  else
	    {
	      unixtime = dos_to_unix_time ((unsigned long) FileTime);
	      if (unixtime <= statbuf.st_mtime);
	      {
		free (destination);
		return 1;
	      }
	    }
	}
      //okay put it in destination.
      if (dump_file (unrar, destination, data, data_size) < 0)
	{
	  free (data);
	  free (destination);
	  return -1;
	}
      free (data);
      free (destination);
    }
  return 0;
}

int
unrar_extract (struct unrar_arguments_t *unrar, int num_files, char **files)
{
  ArchiveList_struct list;
  ArchiveList_struct *listptr;
  int i, j, n;
  int retval;
  char *action;
  char *status;
  char *orig_name;
  int num_failed = 0;

  printf ("\n");
  printf ("Extracting from %s\n", unrar->archive_filename);
  printf ("\n");

  list.next = NULL;
  n = urarlib_list (unrar->archive_filename, &list.next);
  if (n < 0)
    return n;

  if (unrar->verbose)
    {
      printf ("showing %d files...\n", n);
      printf ("given a list of %d files.\n", num_files);
    }

  listptr = list.next;
  for (i = 0; i < n; i++, listptr = listptr->next)
    {
      orig_name = strdup (listptr->item.Name);

      replace_backslash_with_slash (listptr->item.Name);
      if (num_files)
	{
	  int found = 0;
	  for (j = 0; j < num_files; j++)
	    {
	      if (unrar->verbose)
		printf ("comparing '%s' vs '%s'\n", files[j],
			listptr->item.Name);
	      if (strcmp (files[j], listptr->item.Name) == 0)
		{
		  found = 1;
		  break;
		}

	    }
	  if (!found)
	    {
	      free (orig_name);
	      if (listptr->item.FileAttr & 0x10)
		continue;
	      action = "Skipping";
	      status = "";
	      show_status_line (action, listptr->item.Name, status);
	      continue;
	    }
	}

      status = "OK";
      if (listptr->item.FileAttr & 0x10)	//is directory.
	{
	  action = "Creating";
	  if ((retval =
	       unrar_extract_directory (unrar, listptr->item.Name)) < 0)
	    {
	      status = "Failed";
	    }
	  else if (retval == 0)
	    {
	      free (orig_name);
	      continue;
	    }

	}
      else			//is file.
	{
	  action = "Extracting";
	  if ((retval =
	       unrar_extract_file (unrar, listptr->item.Name, orig_name,
				   listptr->item.FileTime)) < 0)
	    {
	      status = "Failed";
	    }
	  else if (retval == 1)
	    {
	      action = "Skipping";
	      status = "";
	      show_status_line (action, listptr->item.Name, status);
	      free (orig_name);
	      continue;
	    }
	  else if (retval == 0)
	    {
	      status = "OK";
              show_status_line (action, listptr->item.Name, status);
	      free (orig_name);
	      continue;
	    }
	}
      free (orig_name);
      show_status_line (action, listptr->item.Name, status);
      if (strcmp (status, "Failed") == 0)
	num_failed++;
      if (listptr == NULL)	//sanity check.
	break;
    }

  urarlib_freelist (list.next);
  if (num_failed) {
    printf ("%d Failed\n", num_failed);
    return -1;
  }
  else
    printf ("All OK\n");
  return 0;
}

int
main (int argc, char **argv)
{
  int retval = 0;
  char **files = NULL;
  int num_files = 0;
  
  if (compat_parse_opts (argc, argv, &arguments) == 0) {
    /* compatible mode success */
  } 
  else if (parse_opts (argc, argv, &arguments) < 0)
    exit (1);

  if (arguments.unrar.verbose)
    {
      printf ("archive name is '%s'\n", arguments.unrar.archive_filename);
      printf ("destination directory is '%s'\n",
	      arguments.unrar.destination_dir);
      printf ("mode = %d\n", arguments.unrar.mode);
      printf ("force = %d\n", arguments.unrar.force);
      printf ("extract_newer = %d\n", arguments.unrar.extract_newer);
      printf ("junk_paths = %d\n", arguments.unrar.junk_paths);
    }

  if (arguments.arraylen > 1)
    {
      files = &arguments.args[1];
      num_files = arguments.arraylen - 1;
    }

  show_copyright ();
  if (arguments.unrar.mode == MODE_LIST)
    retval = unrar_list (&arguments.unrar, num_files, files);
  else if (arguments.unrar.mode == MODE_EXTRACT)
    retval = unrar_extract (&arguments.unrar, num_files, files);

  if (retval < 0)
    exit (1);

  exit (0);
}


src/opts.c
/* Copyright (C) 2004  Jeroen Dekkers <jeroen@dekkers.cx>
   Copyright (C) 2004  Ben Asselstine <benasselstine@canada.com>

   This program is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License
   as published by the Free Software Foundation; either version 2
   of the License, or (at your option) any later version.
   
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License along
   with this program; if not, write to the Free Software Foundation, Inc.,
   51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <error.h>
#include <argp.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include "opts.h"

#define FULL_VERSION PACKAGE_NAME " " PACKAGE_VERSION
const char *argp_program_version = FULL_VERSION;
const char *argp_program_bug_address = "<" PACKAGE_BUGREPORT ">";
static char doc[] = "Extract files from rar archives.";
static char args_doc[] = "ARCHIVE [FILE...] [DESTINATION]";
struct arguments_t arguments;

static struct argp_option options[] = {
  {"extract", CMD_EXTRACT, 0, 0, "Extract files from archive (default)", 0},
  {"list", CMD_LIST, 0, 0, "List files in archive", 1},
  {"force", OPT_FORCE, 0, 0, "Overwrite files when extracting", 2},
  {"extract-newer", OPT_EXTRACT_NEWER, 0, 0,
   "Only extract newer files from the archive", 3},
  {"extract-no-paths", OPT_JUNK_PATHS, 0, 0,
   "Don't create directories while extracting", 3},
  {"password", OPT_PASSWORD, 0, 0, "Decrypt archive using a password", 3},
  {"verbose", OPT_VERBOSE, 0, OPTION_HIDDEN, "Verbosely list files extracted", 3},
  {0}
};


void
free_arguments ()
{
  if (arguments.args)
    free (arguments.args);
  if (arguments.unrar.destination_dir)
    free (arguments.unrar.destination_dir);
  if (arguments.unrar.archive_filename)
    free (arguments.unrar.archive_filename);
  if (arguments.unrar.password)
    free (arguments.unrar.password);
  return;
}

static error_t
parse_opt (int key, char *arg, struct argp_state *state)
{
  /* Get the INPUT argument from `argp_parse', which we
     know is a pointer to our arguments structure. */

  struct arguments_t *arguments = state->input;
  char *pass;

  switch (key)
    {
    case CMD_EXTRACT:
      arguments->unrar.mode = MODE_EXTRACT;
      break;
    case CMD_LIST:
      arguments->unrar.mode = MODE_LIST;
      break;
    case OPT_FORCE:
      arguments->unrar.force = 1;
      break;
    case OPT_JUNK_PATHS:
      arguments->unrar.junk_paths = 1;
      break;
    case OPT_EXTRACT_NEWER:
      arguments->unrar.extract_newer = 1;
      break;
    case OPT_PASSWORD:
      if (arguments->unrar.password == NULL)
      {
	      if ((pass = getpass ("Password:")))
		arguments->unrar.password = strdup (pass);
      }
      break;
    case OPT_VERBOSE:
      arguments->unrar.verbose = 1;
      break;
    case ARGP_KEY_INIT:
      free_arguments ();
      arguments->args = NULL;
      arguments->arraylen = 0;
      break;
    case ARGP_KEY_ARG:
      arguments->arraylen++;
      arguments->args =
	(char **) realloc (arguments->args,
			   arguments->arraylen * sizeof (char *));
      arguments->args[state->arg_num] = arg;
      break;
    case ARGP_KEY_END:
      if (arguments->arraylen > 0)
	{
	  char *file;
	  struct stat statbuf;
	  file = realpath (arguments->args[0], NULL);
	  if (lstat (file, &statbuf) == -1)
	    {
	      free (file);
	      error (0, 0, "invalid archive '%s': %m", arguments->args[0]);
	      argp_usage (state);
	    }
	  else
	    {
	      if (!(S_ISDIR (statbuf.st_mode)))
		arguments->unrar.archive_filename = file;
	      else
		{
		  free (file);
		  error (0, 0, "invalid archive '%s': is a directory",
			 arguments->args[0]);
		  argp_usage (state);
		}
	    }
	}

      if (arguments->arraylen > 1)
	{
	  char *dir;
	  struct stat statbuf;
	  dir = realpath (arguments->args[arguments->arraylen - 1], NULL);
	  if (lstat (dir, &statbuf) == -1)
	    free (dir);
	  else
	    {
	      if (S_ISDIR (statbuf.st_mode))
		{
		  arguments->unrar.destination_dir = dir;
		  arguments->args[arguments->arraylen - 1] = NULL;
		  arguments->arraylen--;
		}
	      else
		free (dir);
	    }
	}
      break;
    default:
      return ARGP_ERR_UNKNOWN;
    }
  return 0;
}

int
set_default_arguments (struct unrar_arguments_t *unrar)
{
  if (unrar->mode == MODE_UNKNOWN)
    unrar->mode = MODE_EXTRACT;
  if (unrar->destination_dir == NULL)
    unrar->destination_dir = realpath ("./", NULL);
  return 0;
}

struct argp argp = { options, parse_opt, args_doc, doc };

int
parse_opts (int argc, char **argv, struct arguments_t *arguments)
{
  int retval;
  setenv ("ARGP_HELP_FMT", "no-dup-args-note", 1);
  atexit (free_arguments);
  retval = argp_parse (&argp, argc, argv, 0, 0, arguments);
  if (retval < 0)
    argp_help (&argp, stdout, ARGP_HELP_EXIT_ERR | ARGP_HELP_SEE,
	       PACKAGE_NAME);
  set_default_arguments (&arguments->unrar);

  if (arguments->arraylen == 0)
    {
      error (0, 0, "Archive not specified\n");
      argp_help (&argp, stdout, ARGP_HELP_EXIT_ERR | ARGP_HELP_SEE,
		 PACKAGE_NAME);
      return -1;
    }

  return 0;
}


/**
 * @brief Decide a valid unrar command.
 * To decide a string is an unrar command or not
 * @param a the string
 * @retval true it is an unrar command
 * @retval false it is not an unrar command
 */
int compat_iscmd(char *a) {
  if (strcmp(a,"e")==0
   || strcmp(a,"l")==0
   || strcmp(a,"lt")==0
   || strcmp(a,"lb")==0
   || strcmp(a,"p")==0
   || strcmp(a,"t")==0
   || strcmp(a,"v")==0
   || strcmp(a,"vt")==0
   || strcmp(a,"vb")==0
   || strcmp(a,"x")==0
  ) {
    return (1==1);
  }
  return (1==0);
}

/**
 * @brief To execute an unrar command.
 * Set arguments to execute an unrar command
 * @param a the command
 * @param the arguments structure
 */
void compat_execcmd(char *a, struct arguments_t *arguments) {
  if (strcmp(a,"e")==0) {
    arguments->unrar.mode = MODE_EXTRACT;
    arguments->unrar.junk_paths = 1;
  } else if (strcmp(a,"l")==0) {
    arguments->unrar.mode = MODE_LIST;
  } else if (strcmp(a,"lt")==0) {
    arguments->unrar.mode = MODE_LIST;
  } else if (strcmp(a,"lb")==0) {
    arguments->unrar.mode = MODE_LIST;
  } else if (strcmp(a,"p")==0) {
  } else if (strcmp(a,"t")==0) {
    arguments->unrar.mode = MODE_LIST;
  } else if (strcmp(a,"v")==0) {
    arguments->unrar.mode = MODE_LIST;
    arguments->unrar.verbose = 1;
  } else if (strcmp(a,"vt")==0) {
    arguments->unrar.mode = MODE_LIST;
    arguments->unrar.verbose = 1;
  } else if (strcmp(a,"vb")==0) {
    arguments->unrar.mode = MODE_LIST;
    arguments->unrar.verbose = 1;
  } else if (strcmp(a,"x")==0) {
    arguments->unrar.mode = MODE_EXTRACT;
  } else {
    return;
  }
}

/**
 * @brief Execute an unrar switch.
 * Set arguments to execute an unrar switch
 * @param a the string
 * @param arguments the arguments strcture
 * @retval 0 it is a switch (start with "-") and arguments is changed
 * @retval 1 it is not a switch and arguments is unchanged
 * @retval 2 it is a stop switch "--" and arguments is unchanged
 */
int compat_execswitch(char *a, struct arguments_t *arguments) {
  char *swtch,*opt,*pass;
  if (a[0]=='-') {
    swtch=&(a[1]);
  } else {
    /* not a switch */
    return 1;
  }
  
  if (strcmp(swtch,"-")==0) {
    /* stop switch */
    return 2;
  } else if (strcmp(swtch,"ep")==0) {
    arguments->unrar.junk_paths = 1;
  } else if (strcmp(swtch,"o+")==0) {
    arguments->unrar.force = 1;
  } else if (strcmp(swtch,"o-")==0) {
    arguments->unrar.force = 0;
  } else if (strncmp(swtch,"p",1)==0) {
    opt = &(swtch[1]);
    if (strcmp(opt,"-")==0) {
      /* don't ask for password */
    } else if (opt[0] != '\0') {
      /* copy password from cmd line */
      pass = opt;
      arguments->unrar.password = strdup(pass);
    } else {
      /* ask password */
      if ((pass = getpass ("Password:")) != NULL) {
        arguments->unrar.password = strdup(pass);
      }
    }
  } else if (strcmp(swtch,"u")==0) {
    arguments->unrar.extract_newer = 1;
  } else {
  }
  return 0;
}

/**
 * @brief Add an string to arguments.args.
 * Add a string to an arguments structure
 * @param a the string
 * @param arguments the arguments strcture
 */
void compat_addargs(char *a, struct arguments_t *arguments) {
  int arg_num;
  arg_num = arguments->arraylen;
  arguments->arraylen++;
  arguments->args =
    (char **) realloc (arguments->args,
                       arguments->arraylen * sizeof (char *));
    arguments->args[arg_num] = a;
}

int
compat_parse_opts(int argc, char **argv, struct arguments_t *arguments1) {
  int retval=0,i;
  /* iscmd? */
  if (argc <=2 || !compat_iscmd(argv[1])) {
    return -1;
  }
  /* init */
  free_arguments();
  arguments1->args=NULL;
  arguments1->arraylen=0;
  /* cmd */
  compat_execcmd(argv[1],arguments1);
  /* switch */
  for (i=2 ; i<argc ; i++) {
    retval = compat_execswitch(argv[i],arguments1);
    if (retval == 1) {
      break;
    } else if (retval == 2) {
      i++;
      break;
    }
  }
  /* args */
  for ( ; i<argc ; i++) {
    compat_addargs(argv[i],arguments1);
  }
  /* end */
  retval=0;
  if (arguments1->arraylen > 0) {
    char *file;
    struct stat statbuf;
    file = realpath (arguments1->args[0], NULL);
    if (lstat (file, &statbuf) == -1) {
      free (file);
      /* error (0, 0, "invalid archive '%s': %m", arguments1->args[0]); */
      retval=-1;
    } else {
      if (!(S_ISDIR (statbuf.st_mode))) {
        arguments1->unrar.archive_filename = file;
      } else {
        free (file);
        /*error (0, 0, "invalid archive '%s': is a directory",
               arguments1->args[0]);*/
        retval=-1;
      }
    }
  }
  if (retval<0) {
    free_arguments();
    memset(&arguments,0,sizeof(arguments));
    return -1;
  }
  if (arguments1->arraylen > 1) {
    char *dir;
    struct stat statbuf;
    dir = realpath (arguments1->args[arguments1->arraylen - 1], NULL);
    if (lstat (dir, &statbuf) == -1)
      free(dir);
    else {
      if (S_ISDIR (statbuf.st_mode)) {
        arguments1->unrar.destination_dir = dir;
        arguments1->args[arguments1->arraylen - 1] = NULL;
        arguments1->arraylen--;
      } else 
        free (dir);
    }
  }
  
  set_default_arguments (&arguments1->unrar);

  if (arguments1->arraylen == 0)
  {
    /*error (0, 0, "Archive not specified\n");*/
    free_arguments();
    memset(&arguments,0,sizeof(arguments));
    return -1;
  }
  /* argument parsing success, register free_arguments to atexit */
  atexit (free_arguments);
  return 0;
}
